docs: attention update (#183)
bentoml/llm-inference-in-production: docs/llm-inference-basics/how-does-llm-inference-work.md

seen so far. This repeats until a stopping criterion is met.

## What's inside an LLM?

Most modern LLMs are
[decoder-only Transformer models](#are-all-llms-decoder-only-transformer-models).
At a high level, the architecture has three major components:

1. **Input representation**: Token IDs are converted into numerical vectors
   (embeddings). The model also incorporates positional information so it can
   distinguish token order. For example, Rotary Positional Embeddings (RoPE)
   encode position within the attention computation.
2. **A stack of Transformer layers**: The embeddings pass through many
   Transformer layers. A typical layer combines a self-attention mechanism,
   which mixes information across tokens, with a feed-forward network, which
   transforms each token independently.
3. **Output head**: The final hidden state is projected into logits, one raw
   score for every token in the vocabulary. These logits are what the model
   [samples from to pick the next token](#how-are-tokens-selected-via-sampling).

Many inference optimizations target either attention computation or the KV
cache that attention uses during generation.

## The attention mechanism

> [Attention](https://arxiv.org/abs/1706.03762) lets the model assign different

Reviewer: Please remove this link or move it farther down the page as an extra resource (it’s distracting to offer a link at the beginning of a sentence that intends to explain that very thing). For example, `For a more detailed description of the attention mechanism, see the [Attention Is All You Need paper](https://arxiv.org/abs/1706.03762)`.

levels of importance to different tokens, so it can capture relationships
across the sequence. For each token, the model computes three vectors:

- **Query (Q)**: what the current token is looking for
- **Key (K)**: what each token offers for matching

> - **Value (V)**: the content each token contributes

Reviewer: These definitions leave me with more questions. They both look like incomplete sentences. Please elaborate a bit more. One thing that I think also helps is to explain the directional nature of each: The query is a "proactive" query from a token when it wants to learn about neighbor tokens, whereas both the key and value are shared in "response" to such a query.

> The model compares queries against keys to produce attention scores,

Reviewer: I know we’re not trying to explain this process in complete detail, but I’d like to see a bit more explanation here. I think the "comparison" you’re referring to here is the comparison between each token multiplied by the query/key, and a high value means they strongly attend to each other. So maybe say something like this (but please correct me if I’m wrong!):

```suggestion
These three vectors help the attention mechanism decide how
important a given token is in relation to another one—how much
it "attends to" that token (also known as its attention score).
The actual process to calculate the attention score is a lot of 
math, but it essentially multiplies the _query_ vector with the
vector of the "querying" token, and multiplies the _key_
vector with the vector of the "responding" token. That
produces two new vectors, which are then multiplied together
to get the attention score. The attention mechanism then

```

Author: I just realized this is not what I really mean. 

> It essentially multiplies the query vector with the vector of the querying token...

This is not accurate. The query vector is already computed from the querying token's hidden representation. You do need to multiply the query vector with the querying token again.

> multiplies the _key_ vector with the vector of the "responding" token

same issue. The key vector is already derived from the responding token.

I guess you mean the learned Q/K/V projection matrices with the resulting query, key, and value vectors. There are three weight matrices (W_Q, W_K, W_V), which are shared across positions. Each position’s hidden state is multiplied by these matrices to produce the query, key, and value vectors of that position. The query vector is therefore the output of the projection, not one of inputs. (I think this is what you mean?)

btw, I just added the three weight matrices before explaining Q, K, and V. A reader who's new to Transformers is likely to wonder, "where did these vectors come from?" before they get the answer.

normalizes the scores with softmax into weights, and uses those weights to
take a weighted sum of the values. In self-attention, queries, keys, and values

> come from the same sequence. A causal mask limits each token to the current and

Reviewer: After explaining how the attention score is derived from the query and key, this paragraph is missing an explanation of how the _value_ vector is used. I think something short here should actually end the paragraph. But I’m sincerely confused by the sentence, "In self-attention, queries, keys, and values
come from the same sequence." I feel like this is trying to clarify something that’s not realistically going to be confused for the target reader here, so I think you can drop it.

For example:

```suggestion
normalizes the scores with softmax so we have a cleanly
distributed ranking of how important every "responding" token
is to the "querying" token. Finally, the _value_ vector for each
"responding" token is multiplied by the vector of the
"querying" token. With some more math, this value is
now embedded into the "querying" token. That means
the token _absorbs_ some high-dimensional values
from every other token in the sequence, relative to how
important each one is (how large its attention score is).

A causal mask limits each token to the current and
```

Author: > Finally, the _value_ vector for each "responding" token is multiplied by the vector of the "querying" token.

I believe each value vector is multiplied by its attention weight, not by the query. The query is no longer involved once the attention weights have been computed.

I removed the self-attention stuff and just focus on how attention scores are converted into updated token representation

1. **Input representation**: Token IDs are converted into numerical vectors
   (embeddings). The model also incorporates positional information so it can
   distinguish token order. For example, Rotary Positional Embeddings (RoPE)
   encode position within the attention computation.
2. **A stack of Transformer layers**: The embeddings pass through many
   Transformer layers. A typical layer combines a self-attention mechanism,
   which mixes information across tokens, with a feed-forward network, which
   transforms each token independently.
3. **Output head**: The final hidden state is projected into logits, one raw
   score for every token in the vocabulary. These logits are what the model
   [samples from to pick the next token](#how-are-tokens-selected-via-sampling).

Many inference optimizations target either attention computation or the KV
cache that attention uses during generation.

## The attention mechanism

Attention lets the model assign different levels of importance to different
tokens, so it can capture relationships across the sequence.

To do this, each attention layer learns three weight matrices: $W_Q$, $W_K$, and

> $W_V$. These matrices are shared across all token positions and transform the

Reviewer: I like the added expressions here, but it’s not clear how they correspond to the three list items that follow.  I think a simple parentheses for each will help.

```suggestion
To do this, each attention layer learns three weight matrices:
$W_Q$ (query weight), $W_K$ (key weight), and $W_V$ (value weight).
These matrices are shared across all token positions and transform the
```

hidden representation of each token into three vectors:

- **Query (Q)**: A proactive request from the current token. It describes what
  information the token wants to gather from positions that it can attend to.
- **Key (K)**: A label describing what information each position has to offer,
  used to judge how well that position matches a given query.
- **Value (V)**: The content each position offers. The position's contribution

>   depends on how well it matches relative to the other positions in view.

Reviewer: These are all high-dimensional tensors, so the first sentence in the definition should to be clear that these are just representations, not a literal "request" or "label"—they’re not string values. It’s an ambiguous concept to explain, but I also get confused by your use of "position," so I think this is more clear:

```suggestion
- **Query (Q)**: A vector that represents a type of information
  the token wants to collect from the other tokens it can attend to.
- **Key (K)**: A vector that represents the type of information the
  token offers. If the key vector is positively aligned with another
  token’s query vector, that other token will "attend to" this token
  (to a degree specified by an attention score, calculated based on
  how well the query and key vectors align).
- **Value (V)**: A vector that represents the actual information the
  token offers to other tokens. If another token attends to this token,
  it will embed a proportion of this value into its token embedding
  (as determined by the attention score).
```
